Index: fuse-emulator/fuse/AUTHORS
===================================================================
--- fuse-emulator/fuse/AUTHORS	(revision 4108)
+++ fuse-emulator/fuse/AUTHORS	(working copy)
@@ -48,6 +48,8 @@
 
 * Michael D Wynne: the Opus Discovery emulation (for EightyOne).
 
+* Patrik Persson: the SpeccyBoot Ethernet emulation.
+
 * Philip Kendall: everything else.
 
 $Id$
Index: fuse-emulator/fuse/man/fuse.1
===================================================================
--- fuse-emulator/fuse/man/fuse.1	(revision 4108)
+++ fuse-emulator/fuse/man/fuse.1	(working copy)
@@ -618,6 +618,8 @@
 .I "\-\-rom\-plusd file"
 .br
 .I "\-\-rom\-beta128 file"
+.br
+.I "\-\-rom\-speccyboot file"
 .RS
 Specify the file to be used for ROM(s) used for each machine. The
 options respectively refer to the 16K Spectrum
@@ -650,8 +652,10 @@
 .RI ( opus.rom ),
 the +D ROM
 .RI ( plusd.rom ),
-and the TR-DOS ROM for Beta 128 emulation with the 48K, TC2048, 128K or +2
-.RI ( trdos.rom ).
+the TR-DOS ROM for Beta 128 emulation with the 48K, TC2048, 128K or +2
+.RI ( trdos.rom ),
+and the SpeccyBoot ROM
+.RI ( speccyboot.rom ).
 The names in brackets denote the defaults.
 .RE
 .PP
@@ -759,6 +763,18 @@
 option.
 .RE
 .PP
+.I "\-\-speccyboot"
+.RS
+Emulate a SpeccyBoot Ethernet interface. Same as the Peripherals Options dialog's
+.I "SpeccyBoot"
+option.
+.RE
+.PP
+.I "\-\-speccyboot-tap device"
+.RS
+Specify the TAP device to use for SpeccyBoot emulation.
+.RE
+.PP
 .I "\-\-volume\-ay volume"
 .RS
 Sets the relative volume of the AY-3-8912 chip from a range of
Index: fuse-emulator/fuse/fuse.c
===================================================================
--- fuse-emulator/fuse/fuse.c	(revision 4108)
+++ fuse-emulator/fuse/fuse.c	(working copy)
@@ -78,6 +78,7 @@
 #include "slt.h"
 #include "snapshot.h"
 #include "sound.h"
+#include "speccyboot.h"
 #include "spectrum.h"
 #include "tape.h"
 #include "timer/timer.h"
@@ -304,6 +305,7 @@
   if( kempmouse_init() ) return 1;
   if( fuller_init() ) return 1;
   if( melodik_init() ) return 1;
+  if( speccyboot_init() ) return 1;
 
   error = pokefinder_clear(); if( error ) return error;
 
Index: fuse-emulator/fuse/menu.c
===================================================================
--- fuse-emulator/fuse/menu.c	(revision 4108)
+++ fuse-emulator/fuse/menu.c	(working copy)
@@ -221,6 +221,7 @@
   case 16: menu_select_roms_with_title( "Interface I",     40, 1 ); return;
   case 17: menu_select_roms_with_title( "Beta 128",        41, 1 ); return;
   case 18: menu_select_roms_with_title( "+D",              42, 1 ); return;
+  case 19: menu_select_roms_with_title( "SpeccyBoot",      43, 1 ); return;
 
   }
 
Index: fuse-emulator/fuse/configure.in
===================================================================
--- fuse-emulator/fuse/configure.in	(revision 4108)
+++ fuse-emulator/fuse/configure.in	(working copy)
@@ -573,6 +573,24 @@
 fi
 AC_SUBST(TIMER_LIBADD)
 
+dnl See if Linux TAP devices are supported
+AC_MSG_CHECKING(whether Linux TAP devices are supported)
+ac_save_CPPFLAGS="$CPPFLAGS"
+CPPFLAGS="$CPPFLAGS $LIBSPEC_CFLAGS"
+AC_TRY_COMPILE([
+    #include <linux/if_tun.h>
+    #include <net/if.h>
+    #include <sys/ioctl.h>
+  ],[
+    int test1 = IFF_TAP | IFF_NO_PI;
+    int test2 = TUNSETIFF;
+  ],
+  AC_DEFINE([LINUX_TAP], 1, [Define to 1 if Linux TAP devices are supported.])
+  AC_MSG_RESULT(yes),
+  AC_MSG_RESULT(no))
+
+CPPFLAGS="$ac_save_CPPFLAGS"
+
 dnl Work out which standard routines we're missing
 AC_MSG_CHECKING(which standard routines we're missing)
 missing_routines=''
Index: fuse-emulator/fuse/menu_data.dat
===================================================================
--- fuse-emulator/fuse/menu_data.dat	(revision 4108)
+++ fuse-emulator/fuse/menu_data.dat	(working copy)
@@ -94,6 +94,7 @@
 Options/Select ROMs/Interface _I..., Item,, menu_options_selectroms_select,, 16
 Options/Select ROMs/_Beta 128..., Item,, menu_options_selectroms_select,, 17
 Options/Select ROMs/+_D..., Item,, menu_options_selectroms_select,, 18
+Options/Select ROMs/Specc_yBoot..., Item,, menu_options_selectroms_select,, 19
 
 Options/_Filter..., Item,,, menu_filter_detail
 
Index: fuse-emulator/fuse/settings.dat
===================================================================
--- fuse-emulator/fuse/settings.dat	(revision 4108)
+++ fuse-emulator/fuse/settings.dat	(working copy)
@@ -71,6 +71,7 @@
 unittests, boolean, 0
 fuller, boolean, 0
 melodik, boolean, 0
+speccyboot, boolean, 0
 
 sound_device, string, NULL, 'd'
 sound, boolean, 1
@@ -166,6 +167,8 @@
 
 start_scaler_mode, string, "normal", 'g', graphics-filter
 
+speccyboot_tap, string, "tap0",
+
 rom_16, string, "48.rom",
 rom_48, string, "48.rom",
 rom_128_0, string, "128-0.rom",
@@ -206,6 +209,7 @@
 rom_scorpion_3, string, "256s-3.rom",
 rom_spec_se_0, string, "se-0.rom",
 rom_spec_se_1, string, "se-1.rom",
+rom_speccyboot, string, "speccyboot.rom",
 rom_interface_i, string, "if1-2.rom",
 rom_opus, string, "opus.rom",
 rom_plusd, string, "plusd.rom",
Index: fuse-emulator/fuse/periph.c
===================================================================
--- fuse-emulator/fuse/periph.c	(revision 4108)
+++ fuse-emulator/fuse/periph.c	(working copy)
@@ -40,6 +40,7 @@
 #include "joystick.h"
 #include "kempmouse.h"
 #include "melodik.h"
+#include "speccyboot.h"
 #include "periph.h"
 #include "rzx.h"
 #include "settings.h"
@@ -343,6 +344,12 @@
 /* Is the Melodik currently active */
 int periph_melodik_active;
 
+/* What sort of SpeccyBoot does the current machine have */
+periph_present speccyboot_present;
+
+/* Is the SpeccyBoot active */
+int periph_speccyboot_active;
+
 int
 periph_setup( const periph_t *peripherals_list, size_t n )
 {
@@ -362,6 +369,8 @@
 
   periph_register_n( kempmouse_peripherals, kempmouse_peripherals_count );
 
+  periph_register_n( speccyboot_peripherals, speccyboot_peripherals_count );
+
   error = periph_register_n( peripherals_list, n ); if( error ) return error;
 
   kempston_present = PERIPH_PRESENT_NEVER;
@@ -372,6 +381,7 @@
   opus_present = PERIPH_PRESENT_NEVER;
   fuller_present = PERIPH_PRESENT_NEVER;
   melodik_present = PERIPH_PRESENT_NEVER;
+  speccyboot_present = PERIPH_PRESENT_NEVER;
 
   return 0;
 }
@@ -421,6 +431,11 @@
   melodik_present = present;
 }
 
+void
+periph_setup_speccyboot( periph_present present ) {
+  speccyboot_present = present;
+}
+
 static void
 update_cartridge_menu( void )
 {
@@ -543,6 +558,13 @@
     periph_register_n( melodik_peripherals, melodik_peripherals_count );
   }
 
+  switch (speccyboot_present ) {
+  case PERIPH_PRESENT_NEVER: periph_speccyboot_active = 0; break;
+  case PERIPH_PRESENT_OPTIONAL:
+    periph_speccyboot_active = settings_current.speccyboot; break;
+  case PERIPH_PRESENT_ALWAYS: periph_speccyboot_active = 1; break;
+  }
+
   update_cartridge_menu();
   update_ide_menu();
   if1_update_menu();
Index: fuse-emulator/fuse/periph.h
===================================================================
--- fuse-emulator/fuse/periph.h	(revision 4108)
+++ fuse-emulator/fuse/periph.h	(working copy)
@@ -107,6 +107,9 @@
 /* Is the Melodik active */
 extern int periph_melodik_active;
 
+/* Is the SpeccyBoot active */
+extern int periph_speccyboot_active;
+
 int periph_setup( const periph_t *peripherals_list, size_t n );
 void periph_setup_kempston( periph_present present );
 void periph_setup_interface1( periph_present present );
@@ -116,6 +119,7 @@
 void periph_setup_opus( periph_present present );
 void periph_setup_fuller( periph_present present );
 void periph_setup_melodik( periph_present present );
+void periph_setup_speccyboot( periph_present present );
 void periph_update( void );
 
 void periph_register_beta128( void );
Index: fuse-emulator/fuse/Makefile.am
===================================================================
--- fuse-emulator/fuse/Makefile.am	(revision 4108)
+++ fuse-emulator/fuse/Makefile.am	(working copy)
@@ -89,6 +89,7 @@
 	slt.c \
 	snapshot.c \
 	sound.cpp \
+	speccyboot.cpp \
 	spectrum.c \
 	tape.c \
 	ui.c \
@@ -199,6 +200,7 @@
 	slt.h \
 	snapshot.h \
 	sound.h \
+	speccyboot.h \
 	spectrum.h \
 	tape.h \
 	utils.h \
Index: fuse-emulator/fuse/settings.pl
===================================================================
--- fuse-emulator/fuse/settings.pl	(revision 4108)
+++ fuse-emulator/fuse/settings.pl	(working copy)
@@ -760,6 +760,7 @@
   case 40: return &( settings->rom_interface_i );
   case 41: return &( settings->rom_beta128 );
   case 42: return &( settings->rom_plusd );
+  case 43: return &( settings->rom_speccyboot );
   default: return NULL;
   }
 }
Index: fuse-emulator/fuse/ui/options.dat
===================================================================
--- fuse-emulator/fuse/ui/options.dat	(revision 4108)
+++ fuse-emulator/fuse/ui/options.dat	(working copy)
@@ -48,6 +48,7 @@
 Checkbox, +D i(n)terface, plusd, INPUT_KEY_n
 Checkbox, (B)eta 128 interface, beta128, INPUT_KEY_b
 Checkbox, Opus Discovery in(t)erface, opus, INPUT_KEY_t
+Checkbox, S(p)eccyBoot interface, speccyboot, INPUT_KEY_p
 Posthook, periph_update
 
 rzx
